### PR TITLE
Provide templates and assets

### LINKED ISSUE #3520: Introduce templates
## Wanted

Implement template management 
- REST (CRUD)
- Model 
- Persistenz (DB)
- add to open api model
- Documentation

### LINKED ISSUE #3522: Map templates to projects
## Wanted

Implement mapping
- REST (CRUD)
- Model 
- Persistenz (DB)
- add to open api model
- Documentation

### LINKED ISSUE #3523: Template and asset handling on SecHub jobs
## Wanted

SecHub template and asset handling on starting jobs

- [x] introduce template data resolver in commons (for SecHub server, PDS side and also wrpapers) 
- [x] create and add PDS parameter `pds.config.template.metadata.list` which contains information
    about assets for template types and also variable data
- [x] an integration ensures that the `PDS` will receive `pds.config.template.metadata.list` parameter
- [x] ensure asset available for PDS jobs
      (upload in asset storage if not already there, or if checksum is not as expected -
       as a fallback for a db restore or storage problems)

### LINKED ISSUE #3524: PDS asset handling
## Wanted

PDS asset handling

- [x] download asset files given by PDS parameter from storage (S3/NFS)
- [x] validate checksum correct (defined in PDS parameter)
- [x] extract files (for development we also provide possibility to define single files in 
      PDS parameters, but SecHub at production time will always provide only 
      "$pdsProductIdentifier.zip")
- [x] provide ENV variable `PDS_JOB_EXTRACTED_ASSETS_FOLDER` which can be used by 
      launcher scripts.
- [x] write integration test with PDS (change existing one for sechub template and asset handling

### LINKED ISSUE #3525: Final update templates and asset documentation
## Situation

In 
- #3518

the initial concept was created

The initial concept can change while implementing all steps defined inside
- #3521

## Wanted
Update the documentation (architecture, concept and techdoc) after all parts are implemented

### LINKED ISSUE #3553: Introduce assets + REST interfaces + asset storage
Introduce assets 

- [x] New usescases

  - admin upload asset file
  - admin deletes asset file
  - admin deletes complete asset
  - admin fetches asset id list
  - admin fetches asset details
  - admin downloads asset file

- [x] implement logic for asset CRUD operations
- [x] create REST endpoints

- [x] Storage
  - assets shall be stored inside database
  - provide AssetStorage
  - provide possibility to delete part inside storage

- [x] Add actions for assets inside DAUI

- [x] Update openapi file
- [x] Write REST doc tests
- [x] Update documentation

### LINKED ISSUE #3705: Fix flaky test inside SecHubJobRepositoryDBTest
## Situation

Sometimes the test
`nextCanceledOrEndedJobsWithEncryptionPoolIdLowerThan_randomization_works`

fails.

## Analyze
The test checks that after some time every entry will be retrieved by random picking.
The problem here is that normally the queue/source for randomness does reduce over time - but here not directly

Also the test has got a fixed size of tries (30) which is the problem when it comes to randomness -> after 30 times (without changing the encryption pool entry...) there can
be still missing parts.

The test does also not really test for randomness currently: If the method to test would just fetch jobs in a serial order the test logic would not fail!

## Wanted
Change the tests, so it is no longer flaky. But improve it also.

## Solution
Change the test and check that at least one job has been fetched twice while another one is fetched one time. This can be checked fast and should always be valid (if there is a least a bit of randomness).

Additionally there should be tests added for case that
`nextCanceledOrEndedJobsWithEncryptionPoolIdLowerThan` does not find jobs (because there are only newer /same encryption pool ids set) . This is currently not tested.

### PR DESCRIPTION
**Pull request by de-jcup**
This PR

- [x] closes #3520
- [x] closes #3522
- [x] closes #3553
- [x] closes #3523
- [x] closes #3524
- [x] closes #3525
- [x] closes #3705

### CODE REVIEW

**Review by winzj — CHANGES_REQUESTED** [review]
Looks very good, I just found some small parts to change.

Please let me know if I made any false assumptions in my propositions somewhere.

File: sechub-administration/src/main/java/com/mercedesbenz/sechub/domain/administration/project/ProjectTemplateService.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        LOG.info("Unassigned template '{}' from project '{}'", templateId, projectId);
+
+        LOG.debug("Project '{}' has now following templates: {}", templateId, projectId, templates);
+
+    }
+
+    @IsSendingSyncMessage(MessageID.REQUEST_ASSIGN_TEMPLATE_TO_PROJECT)
+    private SecHubProjectTemplates sendAssignRequestAndFetchResult(String templateId, String projectId) {
+
+        DomainMessage message = new DomainMessage(MessageID.REQUEST_ASSIGN_TEMPLATE_TO_PROJECT);
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-administration/src/main/java/com/mercedesbenz/sechub/domain/administration/project/ProjectTemplateService.java:104]: As descussed, this service lacks also tests - currently only tested by integration tests -  will add this if it is not too complicated.

File: sechub-doc/src/docs/asciidoc/documents/shared/snippet/pds-param-template-metadata-syntax.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+  [ //<1>
+    {
+        "templateId" : "templateId", //<2>
+        "tempplateType": "WEBSCAN_LOGIN", //<3>
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/snippet/pds-param-template-metadata-syntax.json:4]: Will add a unit test as well for the pds parameter -> so examples are correct...

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/template/TemplateRestController.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    TemplateService templateService;
+
+    @Autowired
+    AuditLogService auditLogService;
+
+    @Autowired
+    LogSanitizer logSanitizer;
+
+    @UseCaseAdminCreatesOrUpdatesTemplate(@Step(number = 1, next = 2, name = "REST API call to create or update template", needsRestDoc = true))
+    @RequestMapping(path = "/template/{templateId}", method = RequestMethod.POST)
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/template/TemplateRestController.java:48]: You are absolutely right:

https://www.baeldung.com/rest-http-put-vs-post

> Conversely, the PUT method should be used to request that the server store the enclosed entity under the provided Request-URI. If the Request-URI points to an existing resource on the server, the supplied entity will be considered a modified version of the existing resource. Therefore, the PUT method call will either create a new resource or update an existing one

will change it

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/asset/AssetService.java [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+    private UserInputAssertion inputAssertion;
+
+    private CheckSumSupport checkSumSupport;
+
+    private StorageService storageService;
+
+    /* @formatter:off */
+    AssetService(
+            @Autowired AssetFileRepository repository,
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/asset/AssetService.java:57]: Yes, having only one constructor, spring should handle this without annotation at all.

File: sechub-scan/src/test/java/com/mercedesbenz/sechub/domain/scan/asset/AssetFileInformationTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        /* prepare */
+        String sameChecksum = "checksum_dummy";
+
+        AssetFileData info1 = new AssetFileData();
+        info1.setChecksum(sameChecksum);
+        info1.setFileName("filename1");
+
+        AssetFileData info2 = new AssetFileData();
+        info1.setChecksum(sameChecksum);
+        info2.setFileName("filename1");
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-scan/src/test/java/com/mercedesbenz/sechub/domain/scan/asset/AssetFileInformationTest.java:45]: hm.. maybe the test does also not check this correctly.

File: sechub-administration/src/main/java/com/mercedesbenz/sechub/domain/administration/AdministrationAPIConstants.java [hunk 6]
```diff
@@ -108,6 +108,12 @@ private AdministrationAPIConstants() {
 
     public static final String API_CHANGE_PROJECT_ACCESSLEVEL = API_ADMINISTRATION + "project/{projectId}/accesslevel/{projectAccessLevel}";
 
+    /* +-----------------------------------------------------------------------+ */
+    /* +............................ Templates.................................+ */
+    /* +-----------------------------------------------------------------------+ */
+    public static final String API_ASSIGN_TEMPLATE_TO_PROJECT = API_ADMINISTRATION + "project/{projectId}/template/{templateId}";
+    public static final String API_UNASSIGN_TEMPLATE_FROM_PROJECT = API_ADMINISTRATION + "project/{projectId}/template/{templateId}";
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-administration/src/main/java/com/mercedesbenz/sechub/domain/administration/AdministrationAPIConstants.java:115]: We did the same for 
- API_SHOW_PROJECT_DETAILS
- API_CHANGE_PROJECT_DETAILS
- API_DELETE_PROJECT
- ... more

For readability/clearness I would keep it as is.

File: sechub-administration/src/main/java/com/mercedesbenz/sechub/domain/administration/project/ProjectDetailInformation.java [hunk 7]
```diff
@@ -17,16 +17,22 @@ public class ProjectDetailInformation {
     public static final String PROPERTY_OWNER = "owner";
     public static final String PROPERTY_ACCESSLEVEL = "accessLevel";
     public static final String PROPERTY_DESCRIPTION = "description";
+    public static final String PROPERTY_TEMPLATES = "templates";
 
     private String projectId;
 
     private List<String> users = new ArrayList<>();
     private List<String> whitelist = new ArrayList<>();
+    private List<String> templates = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-administration/src/main/java/com/mercedesbenz/sechub/domain/administration/project/ProjectDetailInformation.java:26]: Yes I will change this. But inside Project.java we had also "templates". I will change it there as well to "templateIds"

[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-administration/src/main/java/com/mercedesbenz/sechub/domain/administration/project/ProjectDetailInformation.java:26]: Will change this. Will also change Project.java (an entity) + column names + flyway parts

File: sechub-openapi-java/src/main/resources/openapi.yaml [hunk 8]
```diff
@@ -3789,4 +3915,256 @@ paths:
...
+        "200":
+          description: "Ok"
+          
+        "404":
+          description: "Not found"
+      tags:
+        - Configuration
+        
+  /api/admin/asset/{assetId}/file:
+    post:
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-openapi-java/src/main/resources/openapi.yaml:4074]: This is not the template change mechanism mentioned before.

I will keep this is as - reason: we use POST for every upload mechanism inside SecHub and PDS.

**Pull request merged**